src/backend/distributed/executor/multi_utility.c

```diff
@@ -461,6 +462,91 @@ multi_ProcessUtility(PlannedStmt *pstmt,
 							params, dest, completionTag);
 #endif
 
+#if (PG_VERSION_NUM >= 100000)
+	if (IsA(parsetree, CreateStmt))
+	{
+		CreateStmt *createStatement = (CreateStmt *) parsetree;
+
+		if (createStatement->inhRelations != NIL && createStatement->partbound != NULL)
+		{
+			RangeVar *parentRelation = linitial(createStatement->inhRelations);
+			bool parentMissingOk = false;
+			Oid parentRelationId = RangeVarGetRelid(parentRelation, NoLock,
+													parentMissingOk);
+
+			Assert(parentRelationId != InvalidOid);
+
+			/*
+			 * If a partition is being created and if its parent is a distributed
+			 * table, we will distribute this table as well.
+			 */
+			if (IsDistributedTable(parentRelationId))
+			{
+				bool missingOk = false;
+				Oid relationId = RangeVarGetRelid(createStatement->relation, NoLock,
+												  missingOk);
+				Var *parentDistributionColumn = DistPartitionKey(parentRelationId);
+				char parentDistributionMethod = DISTRIBUTE_BY_HASH;
+				char *parentRelationName = get_rel_name(parentRelationId);
+				bool viaDeprecatedAPI = false;
+
+				CreateDistributedTable(relationId, parentDistributionColumn,
+									   parentDistributionMethod, parentRelationName,
+									   viaDeprecatedAPI);
+			}
+		}
+	}
+	else if (IsA(parsetree, AlterTableStmt))
+	{
+		AlterTableStmt *alterTableStatement = (AlterTableStmt *) parsetree;
+		List *commandList = alterTableStatement->cmds;
+		ListCell *commandCell = NULL;
+
+		foreach(commandCell, commandList)
+		{
+			AlterTableCmd *alterTableCommand = (AlterTableCmd *) lfirst(commandCell);
+
+			if (alterTableCommand->subtype == AT_AttachPartition)
+			{
+				Oid relationId = AlterTableLookupRelation(alterTableStatement, NoLock);
+				PartitionCmd *partitionCommand = (PartitionCmd *) alterTableCommand->def;
+				bool partitionMissingOk = false;
+				Oid partitionRelationId = RangeVarGetRelid(partitionCommand->name, NoLock,
+														   partitionMissingOk);
+
+				/*
+				 * If user first distributes the table then tries to attach it to non
+				 * distributed table, we error out.
+				 */
+				if (!IsDistributedTable(relationId) &&
```

Comment:
I added very detailed comments to explain all four cases and how we handle them. We cannot put an Assert for distributed/distributed case, because related checks in later stage of codepath.